### ISSUE TITLE
Auto connect

### ISSUE THREAD

**Issue by bjonnh** [rppicomidi/midi2piousbhub#3]
I was wondering if you would be interested in having an autoconnect option for BT.

What I was thinking was:

- have a json key `autoconnect` that contains a single string with the address of a server
- if autoconnect is defined, go in client mode, scan and try to find that server at boot time. 
- On disconnect, try connecting again after a delay? (could be an option)
- add functions to add/remove the autoconnect key in the json

My use case is that I would like to have a set of devices always connected when I turn them on.

**Comment by rppicomidi** [comment]
@bjonnh Yes, a similar feature is on my roadmap. Not there yet. Be aware that the Bluetooth MIDI library this project uses only supports a single Bluetooth LE MIDI device connected at one time, so you cannot have a "set" of devices. You can have only one.

Right now, the Bluetooth MIDI starts up in server mode. It means it is waiting for something like an iPad to connect to it. To accomplish what you describe, you need to first have an option to start in client mode instead of server mode, and then you need an option to connect to a particular device, or perhaps the last connected device, on startup.

**Comment by bjonnh** [comment]
Yes that's exactly what I want to happen. If autoconnect is set, do not start in server mode and wait until the given MAC exists then connect to it.

**Comment by rppicomidi** [comment]
@bjonnh What do you want to happen if the Bluetooth device is not there? Do you want it to retry to connect forever? Give up after a while but remain in client mode? Give up after a while and switch to server mode?

**Comment by bjonnh** [comment]
if autoconnect is on it should continue to retry. We could make that optional with a GPIO too. so you could have a switch client/server.
But in my use case, each device will be fixed and used only for one thing.

**Comment by rppicomidi** [comment]
@bjonnh Do you want to have a go at implementing that? I will not be able to get to that feature for a few weeks, I think.

**Comment by bjonnh** [comment]
I can't promise anything in term of timing but yes I'll get to it.

I am also working on a wifi implementation in parallel to see how it behave in term of latency (so far average of 2ms with 99% of values < 4ms)

**Comment by rppicomidi** [comment]
@bjonnh I pushed  WIP code that starts to implement this, but it is buggy. See #5, for example.

**Comment by rppicomidi** [comment]
@bjonnh I believe the latest version of code implements what you want, except perhaps not how you would have implemented it. If it does what you want,  please close this issue.

**Comment by bjonnh** [comment]
I will be able to test that on Wednesday and will let you know. Thanks for all your work.

**Comment by bjonnh** [comment]
So it works if I do that:
- server up
- connect client
- shutdown client
- restart client

and the client will autoconnect.

However if only the server is restarted it doesn't seem to autoreconnect I have to restart the client.

**Comment by rppicomidi** [comment]
Once you connect, use the `save` cli command to make the last connected address part of the preset. Then, on power-up, when the client device restarts, it will continue to attempt to connect until it successfully connects to previously connected server. The reason the code does not auto-save on connect is that saving freezes all code execution until the write to flash completes. That is a limitation of using the boot flash as persistent storage, too.

Once the server is connected once, it won't automatically reconnect, though. That is, if the connection goes away because either the client or server disconnects, then the client won't automatically reconnect. As long as you save the current preset after connection, the code as it is fixes the original use case:

_My use case is that I would like to have a set of devices always connected when I turn them on._

What is the use case for forcing re-connection after the server disconnects? The only case I can think of is the server and client lose connection due to out of range or severe interference.

Automatically reconnecting after the server disconnects could be a annoying if you are trying to connect to the server from a different client. It would be really bad if you forgot to turn this device off and could not get the server to stay disconnected. That said, if you really need the client to connect to the server on power up and to try to stay connected even if the server disconnects, it can probably be done.

**Comment by bjonnh** [comment]
Yes sorry I may have worded that poorly.

It works if the client is freshly booted up, it will auto connect as I had saved before.
But if you turn off the server and put it on again, the client doesn't reconnect. As you have explained.

Yes I was explicitly thinking about interference, moving out of range, turning off instruments (by accident or to move them around during a live session).

**Comment by rppicomidi** [comment]
@bjonnh I added a `keep_client_connected` flag and supporting code in the CLI. I think it does what you want. If you disconnect from the client side, it won't reconnect. But if you disconnect from the server side, it will try to reconnect forever.

**Comment by bjonnh** [comment]
Thanks a lot for all your support it is appreciated.

**Comment by rppicomidi** [comment]
@bjonnh If this code does what you want it to, would you please close this issue or update me on what is not working for you?

**Comment by bjonnh** [comment]
Will do, I will work on it more on Wednesday and update the issue.

**Comment by bjonnh** [comment]
Thanks, that seem to work. Sometimes it needs 5-10s for the client to react, but most of the time it is instant.
